sphericart-torch/include/sphericart/cuda.hpp

```diff
@@ -1,26 +1,26 @@
 #ifndef SPHERICART_TORCH_CUDA_HPP
 #define SPHERICART_TORCH_CUDA_HPP
 
-#include <ATen/Tensor.h>
-
-#include <vector>
-
 namespace sphericart_torch {
 
-bool adjust_cuda_shared_memory(at::ScalarType scalar_type, int64_t l_max,
+template <typename scalar_t>
+void spherical_harmonics_cuda_base(
+    const scalar_t *__restrict__ xyz, const int nedges,
+    const scalar_t *__restrict__ prefactors, const int nprefactors,
+    const int64_t l_max, const bool normalize, const int64_t GRID_DIM_X,
+    const int64_t GRID_DIM_Y, const bool xyz_requires_grad,
+    const bool gradients, const bool hessian, scalar_t *__restrict__ sph,
+    scalar_t *__restrict__ dsph, scalar_t *__restrict__ ddsph);
+
+template <typename scalar_t>
+void spherical_harmonics_backward_cuda_base(
+    const scalar_t *__restrict__ dsph, const scalar_t *__restrict__ sph_grad,
+    const int nedges, const int ntotal, scalar_t *__restrict__ xyz_grad);
+
+bool adjust_cuda_shared_memory(size_t element_size, int64_t l_max,
```

Comment:
this function needs documentation

Reply:
Added documentation in latest commit https://github.com/lab-cosmo/sphericart/commit/9da4642c88c808b6b5e38478e7899f269bc49f03.